src/Microsoft.DotNet.Cli.Utils/ToolCommandName.cs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.DotNet.Cli.Utils
+{
+    public struct ToolCommandName : IEquatable<ToolCommandName>
+    {
+        public ToolCommandName(string name)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
```

Comment:
What do you suggest? I just always throw this exception when I do simple validation on constructor. And by default this exception should only be seen during development. The type should be constructed after the validation of user input(or other boundary of the system). And the validation should be part of the business logic and it is not suitable in constructor.

Reply:
Had a chat with Nick, and for "real" libraries it should throw `ArgumentException` for an empty value.  But since this library is for internal usage it's not really important.

Reply:
Not sure I agree with this statement. Even if this is not a library we are shipping, we should strive to right great code and set good examples. Now, there is a chance that some will come along, see your change like this and actually make a PR with the same change, because that's a pattern we are establishing in the repo. So, we should fix it very soon. And yes, there are other places in the repo where this is true.

Reply:
I agree on internal library should have the same quality as publish library.

However, I don't think this class is a library. On the contrary, this class is a [whole value](http://fit.c2.com/wiki.cgi?WholeValue). Or a domain model. And it should not be shared.

Now on we should throw specific or not. On the cost and benefit point of view, it is very low. The stack trace will point to exact constructor and the dev will know why. And again, if the user see this exception, it is a bug.

Reply:
I think the code should be of as good quality as possible, independently of it being a library (public or internal), an app, a test project or whatever. While the impact of this is super small, so is the fix.

Reply:
The main value of splitting the exception between null and empty case is that you can tell from the exception.ToString() which you hit.

A lot of projects have some sort of validation helper where you can do something like `Requires.NotNullOrEmpty(name, nameof(name))`.

Reply:
Ahhhh, it should be ArgumentException, fixing it

Reply:
https://github.com/dotnet/cli/pull/10001